### PR TITLE
Circular JSON support

### PR DESCRIPTION
**Pull request by shamasis**
Changes:

- all JSON functions replaced by CircularJSON
- the circular-json in `bridge-client.js` is modified (replaced `"\\"` with `String.fromCharCode(92)` to allow being executed as code.
- added / updated tests for circular dependency.

PS: Discovered a bug in `window.postMessage` bleeding data when more than one frames are running. Changed one test for this and took a note to fix the actual issue ASAP in a separate PR.

### CODE REVIEW

File: lib/uvm/bridge-client.js [hunk 1]
```diff
@@ -7,6 +14,9 @@ var STRING = 'string', // constant
...
...u.length),h=this),typeof v=="object"&&v?(a.call(f,v)<0&&f.push(h=v),p=f.length,d=a.call(l,v),d<0?(d=l.push(v)-1,o?(u.push((""+e).replace(s,r)),c[d]=n+u.join(n)):c[d]=c[0]):v=c[d]):typeof v=="string"&&o&&(v=v.replace(r,i).replace(n,r))),v}}function c(e,t){for(var r=0,i=t.length;r<i;e=e[t[r++].replace(o,n)]);return e}function h(e){return function(t,s){var o=typeof s=="string";return o&&s.charAt(0)===n?new f(s.slice(1)):(t===""&&(s=v(s,s,{})),o&&(s=s.replace(u,"$1"+n).replace(i,r)),e?e.call(this,t,s):s)}}function p(e,t,n){for(var r=0,i=t.length;r<i;r++)t[r]=v(e,t[r],n);return t}function d(e,t,n){for(var r in t)t.hasOwnProperty(r)&&(t[r]=v(e,t[r],n));return t}function v(e,t,r){return t instanceof Array?p(e,t,r):t instanceof f?t.length?r.hasOwnProperty(t)?r[t]:r[t]=c(e,t.split(n)):e:t instanceof Object?d(e,t,r):t}function m(t,n,r,i){return e.stringify(t,l(t,n,!i),r)}function g(t,n){return e.parse(t,h(n))}var n="~",r=es+"x"+("0"+n.charCodeAt(0).toString(16)).slice(-2),i=es+r,s=new t(r,"g"),o=new t(i,"g"),u=new t("(?:^|([^"+es+es+"]))"+i),a=[].indexOf||function(e){for(var t=this.length;t--&&this[t]!==e;);return t},f=String;return{stringify:m,parse:g}}(JSON,RegExp,String.fromCharCode(92));
```
Comment (czardoz) [inline lib/uvm/bridge-client.js]: Can we not have the non-minified code here? Browserify will minify it when generating the final file anyway, right?

File: test/unit/uvm-sanity.test.js [hunk 2]
```diff
@@ -32,18 +32,18 @@ describe('uvm', function () {
             var sourceData = 'test',
                 context = uvm.spawn({
                     bootcode: `
-                        bridge.on('loopback', function (data) {
-                            bridge.dispatch('loopback', data);
+                        bridge.on('loopback2', function (data) {
```
Comment (czardoz) [inline test/unit/uvm-sanity.test.js]: Why is this renamed?

Comment (shamasis) [inline test/unit/uvm-sanity.test.js]: That is the whole issue I point out in description @czardoz. In browser mocha, somehow tests are overlapping each other and two parallel sandboxes are sending messages to all hosts due to the lack of target used and id protection. Initially @numaanashraf did mention that even if the chrome docs say otherwise, the behaviour of postMessage is confusing.

So, until we fix this issue, I renamed one of the test.
On a different note, I will rename the new test and nothing will be visible here.

File: lib/uvm/bridge-client.js (hunk 1 above)
Comment (shamasis) [inline lib/uvm/bridge-client.js]: @czardoz I am afraid not. Minification and other things is not UVM's job. The bridge code is like core firmware of UVM. It is run before bootcode is run in sandbox. And as such this is not minified.

The unmagnified code is in Node Modules :-)

File: lib/uvm/bridge-client.js [hunk 3]
```diff
@@ -27,13 +37,13 @@ bridge = (function (emit, jsonStringify, arrayProtoSlice) {
...
+}(__uvm_emit, CircularJSON.stringify, Array.prototype.slice));
 
 __uvm_dispatch = (function (bridge, bridgeEmit, jsonParse) {
     return function (args) {
         bridgeEmit.apply(bridge, jsonParse(args));
     };
-}(bridge, bridge.emit, JSON.parse));
-
+}(bridge, bridge.emit, CircularJSON.parse));
+delete CircularJSON;
```
Comment (vegetableman) [inline lib/uvm/bridge-client.js]: Does it delete the variable from the context? Tried it, doesn't seem so.
screen shot 2016-12-19 at 2 04 05 pm

Comment (vegetableman) [inline lib/uvm/bridge-client.js]: More details on delete:- http://perfectionkills.com/understanding-delete/

Comment (shamasis) [inline lib/uvm/bridge-client.js]: Right. Stupid mistake from my end. Removing the `var` will cause `delete` to work AFAIK.

Comment (shamasis) [inline lib/uvm/bridge-client.js]: @vegetableman done.

**Comment by shamasis** [comment]
any other aspect? @vegetableman @SamvelRaja

**Comment by SamvelRaja** [comment]
@shamasis  Seeing it..

File: lib/uvm/bridge-client.js [hunk 4]
```diff
@@ -7,6 +14,9 @@ var STRING = 'string', // constant
...
...u.length),h=this),typeof v=="object"&&v?(a.call(f,v)<0&&f.push(h=v),p=f.length,d=a.call(l,v),d<0?(d=l.push(v)-1,o?(u.push((""+e).replace(s,r)),c[d]=n+u.join(n)):c[d]=c[0]):v=c[d]):typeof v=="string"&&o&&(v=v.replace(r,i).replace(n,r))),v}}function c(e,t){for(var r=0,i=t.length;r<i;e=e[t[r++].replace(o,n)]);return e}function h(e){return function(t,s){var o=typeof s=="string";return o&&s.charAt(0)===n?new f(s.slice(1)):(t===""&&(s=v(s,s,{})),o&&(s=s.replace(u,"$1"+n).replace(i,r)),e?e.call(this,t,s):s)}}function p(e,t,n){for(var r=0,i=t.length;r<i;r++)t[r]=v(e,t[r],n);return t}function d(e,t,n){for(var r in t)t.hasOwnProperty(r)&&(t[r]=v(e,t[r],n));return t}function v(e,t,r){return t instanceof Array?p(e,t,r):t instanceof f?t.length?r.hasOwnProperty(t)?r[t]:r[t]=c(e,t.split(n)):e:t instanceof Object?d(e,t,r):t}function m(t,n,r,i){return e.stringify(t,l(t,n,!i),r)}function g(t,n){return e.parse(t,h(n))}var n="~",r=es+"x"+("0"+n.charCodeAt(0).toString(16)).slice(-2),i=es+r,s=new t(r,"g"),o=new t(i,"g"),u=new t("(?:^|([^"+es+es+"]))"+i),a=[].indexOf||function(e){for(var t=this.length;t--&&this[t]!==e;);return t},f=String;return{stringify:m,parse:g}}(JSON,RegExp,String.fromCharCode(92));
```
Comment (SamvelRaja) [inline lib/uvm/bridge-client.js]: I wonder why we can't add it as a module and require it here. Am i missing something really big @shamasis ?

File: lib/uvm/bridge-client.js [hunk 5]
```diff
@@ -14,11 +14,15 @@ var STRING = 'string', // constant
...
+(0,eval(\` // we eval it to allow global deletion http://stackoverflow.com/a/25919959
```
Comment (vegetableman) [inline lib/uvm/bridge-client.js]: @shamasis How does `eval` help here?. Wouldn't the below added closure below take of the variable scope.

Comment (vegetableman) [inline lib/uvm/bridge-client.js]: That is, How does it help in the deletion of globals?

Comment (vegetableman) [inline lib/uvm/bridge-client.js]: Both below, do the same thing:
-  `vm.runInContext("var m = 1; (0, eval((function(){var y = 1; m = 3}())))", c)`
-  ` vm.runInContext("var m = 1; (function(){var y = 1; m = 3}())", c)`
